File: project-{topic}-pro/backend/src/main/java/com/alx/taskmgr/controller/TaskController.java
package com.alx.taskmgr.controller;

import com.alx.taskmgr.dto.task.TaskCreateRequest;
import com.alx.taskmgr.dto.task.TaskResponse;
import com.alx.taskmgr.dto.task.TaskUpdateRequest;
import com.alx.taskmgr.service.TaskService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST Controller for managing task operations.
 */
@RestController
@RequestMapping("/api/tasks")
@RequiredArgsConstructor
public class TaskController {

    private final TaskService taskService;

    /**
     * Creates a new task.
     * @param request TaskCreateRequest DTO.
     * @return ResponseEntity with the created task's details.
     */
    @PostMapping
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<TaskResponse> createTask(@Valid @RequestBody TaskCreateRequest request) {
        TaskResponse newTask = taskService.createTask(request);
        return new ResponseEntity<>(newTask, HttpStatus.CREATED);
    }

    /**
     * Retrieves a task by its ID.
     * @param id The ID of the task.
     * @return ResponseEntity with TaskResponse DTO.
     */
    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<TaskResponse> getTaskById(@PathVariable Long id) {
        TaskResponse task = taskService.getTaskById(id);
        return ResponseEntity.ok(task);
    }

    /**
     * Retrieves all tasks for a specific project.
     * @param projectId The ID of the project.
     * @return ResponseEntity with a list of TaskResponse DTOs.
     */
    @GetMapping("/project/{projectId}")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<List<TaskResponse>> getTasksByProjectId(@PathVariable Long projectId) {
        List<TaskResponse> tasks = taskService.getTasksByProjectId(projectId);
        return ResponseEntity.ok(tasks);
    }

    /**
     * Retrieves all tasks assigned to the current authenticated user.
     * @return ResponseEntity with a list of TaskResponse DTOs.
     */
    @GetMapping("/assigned-to-me")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<List<TaskResponse>> getTasksAssignedToCurrentUser() {
        List<TaskResponse> tasks = taskService.getTasksAssignedToCurrentUser();
        return ResponseEntity.ok(tasks);
    }

    /**
     * Updates an existing task.
     * @param id The ID of the task to update.
     * @param request TaskUpdateRequest DTO containing updated details.
     * @return ResponseEntity with the updated task's details.
     */
    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<TaskResponse> updateTask(@PathVariable Long id, @Valid @RequestBody TaskUpdateRequest request) {
        TaskResponse updatedTask = taskService.updateTask(id, request);
        return ResponseEntity.ok(updatedTask);
    }

    /**
     * Deletes a task.
     * @param id The ID of the task to delete.
     * @return ResponseEntity indicating success.
     */
    @DeleteMapping("/{id}")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<Void> deleteTask(@PathVariable Long id) {
        taskService.deleteTask(id);
        return ResponseEntity.noContent().build();
    }
}

File: project-{topic}-plus/src/api/controllers/taskController.js
```javascript
const taskService = require('../../services/taskService');
const catchAsync = require('../../utils/catchAsync');
const logger = require('../../config/logger');

/**
 * Create a new task for the authenticated user.
 */
exports.createTask = catchAsync(async (req, res, next) => {
  const newTask = await taskService.createTask(req.user.id, req.body);
  logger.info(`User ${req.user.id} created task: ${newTask.title}`);

  res.status(201).json({
    status: 'success',
    data: {
      task: newTask,
    },
  });
});

/**
 * Get all tasks for the authenticated user.
 */
exports.getAllTasks = catchAsync(async (req, res, next) => {
  const { tasks, total, results } = await taskService.getAllTasks(req.user.id, req.query);
  logger.debug(`User ${req.user.id} fetched ${results} of ${total} tasks.`);

  res.status(200).json({
    status: 'success',
    results,
    total,
    data: {
      tasks,
    },
  });
});

/**
 * Get a specific task by ID for the authenticated user.
 */
exports.getTaskById = catchAsync(async (req, res, next) => {
  const task = await taskService.getTaskById(req.params.id, req.user.id);
  logger.debug(`User ${req.user.id} fetched task: ${task.id}`);

  res.status(200).json({
    status: 'success',
    data: {
      task,
    },
  });
});

/**
 * Update a task for the authenticated user.
 */
exports.updateTask = catchAsync(async (req, res, next) => {
  const updatedTask = await taskService.updateTask(req.params.id, req.user.id, req.body);
  logger.info(`User ${req.user.id} updated task: ${updatedTask.id}`);

  res.status(200).json({
    status: 'success',
    data: {
      task: updatedTask,
    },
  });
});

/**
 * Delete a task for the authenticated user.
 */
exports.deleteTask = catchAsync(async (req, res, next) => {
  await taskService.deleteTask(req.params.id, req.user.id);
  logger.info(`User ${req.user.id} deleted task: ${req.params.id}`);

  res.status(204).json({
    status: 'success',
    data: null,
  });
});
```